### PR TITLE
issue #131 - adding command palette option to turn dev mode on and off

### PR DESCRIPTION
**Pull request by bfitzpat**
Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

### CODE REVIEW

**Comment by bfitzpat** [comment]
Essentially the thought here is that the --dev mode option can be turned on temporarily in a given session. So the user goes to the command palette and types "Apache Camel K: Enable --dev mode for all future deployments", works on their integration, and deploys it maybe in --configmap mode. It's essentially --configmap and --dev on the command line that we're invoking.

The user should then be able to turn the mode off by going back to the command palette and typing "Apache Camel K: Disable --dev mode..."

File: src/IntegrationUtils.ts [hunk 1]
```diff
@@ -322,6 +325,8 @@ function createNewIntegration(integrationFileUri: vscode.Uri, devMode? : boolean
 				console.error(error); 
 			});
 
+		devMode = extension.getDevMode();
```
Comment (apupier) [inline src/IntegrationUtils.ts]: What is happening when use is using the "Start integration in dev mode" but never "started dev mode" or either "stopped dev mode" previously? it looks alke the command "Start integration in dev mode will not start in dev mode.

Comment (bfitzpat) [reply src/IntegrationUtils.ts]: Taking a look

Comment (bfitzpat) [reply src/IntegrationUtils.ts]: Updated

File: src/IntegrationUtils.ts [hunk 2]
```diff
@@ -322,41 +325,55 @@ function createNewIntegration(integrationFileUri: vscode.Uri, devMode? : boolean
...
 		}			
-		console.log(`commandString = ${commandString}`);
+		console.log(`commandString = ${kamelCLI} ${kamelArgs}`);
 		if (devMode && devMode === true) {
 			if (extension.mainOutputChannel) {
 				extension.mainOutputChannel.show();
 			}
 		}
-		let runKubectl = child_process.exec(commandString, { cwd : foldername});
+		let runKubectl = child_process.spawn(kamelCLI, kamelArgs, { cwd : foldername});
```
Comment (bfitzpat) [inline src/IntegrationUtils.ts]: Using child_process.exec was creating a subprocess that was not getting killed when the main process was killed, so switching to spawn eliminates that issue and we have a much cleaner process that can be killed properly.

Comment (bfitzpat) [reply src/IntegrationUtils.ts]: This was the key - https://stackoverflow.com/questions/27770537/two-processes-created-by-child-process-exec

**Comment by bfitzpat** [comment]
One important note. This is only a per-session setting. It does not persist currently from session to session, so if you shut down VS Code and restart, it starts back at "--dev disabled" mode. I'm not sure if we want to make it a setting or not. Any thoughts? @apupier @lhein ?

**Comment by lhein** [comment]
If we want to persist it then we will have to drop the command and introduce a setting. In my opinion that is the way to go. Maybe we should look at other commands if there are any commands that alter settings. If that is the case we can surely keep the command too.

File: src/IntegrationUtils.ts [hunk 3]
```diff
@@ -322,41 +325,55 @@ function createNewIntegration(integrationFileUri: vscode.Uri, devMode? : boolean
 				console.error(error); 
 			});
 
-		let commandString = `kamel run "${absoluteRoot}"`;
+		if (extension.getDevMode() === true) {
+			// it true, always override whatever the dev mode is that's passed in
```
Comment (apupier) [inline src/IntegrationUtils.ts]: typo comment it -> if

**Comment by bfitzpat** [comment]
Ok. I will spin the exec => spawn change off into a separate change then -- that needs to get in to avoid the hanging process. But will leave the rest for a future PR.

**Comment by bfitzpat** [comment]
Spun off the exec to spawn work in https://github.com/camel-tooling/vscode-camelk/pull/140 and will continue work on the --dev mode stuff in a later release.

**Comment by bfitzpat** [comment]
Closing this one and will revisit in another PR at another date

**Pull request closed without merging**